src/test/regress/input/dispatch.source

```diff
@@ -625,3 +625,97 @@ reset gp_print_create_gang_time;
 reset optimizer;
 drop function cleanupAllGangs();
 drop table t_create_gang_time;
+
+--
+-- Backend process GUCs marked as GUC_GPDB_NEED_SYNC are overwritten by
+-- coordinator upon receiving a new connection.
+--
+-- GUCs that were changed in the configuration file of the QD should be re-sent
+-- to segments on gpstop -u, if GUC_GPDB_NEED_SYNC is present if the flags.
+--
+
+-- start_ignore
+DROP ROLE IF EXISTS dispatch_test_user2;
+DROP TABLE IF EXISTS dispatch_test_session_not_dead;
+-- end_ignore
+
+-- This table shouldn't be dropped until the end of the test.
+CREATE TEMPORARY TABLE dispatch_test_session_not_dead(a int);
+
+-- Use a separate user without any priveleges to catch possible context issues.
+CREATE ROLE dispatch_test_user2 NOSUPERUSER;
+SET SESSION AUTHORIZATION dispatch_test_user2;
+
+CREATE OR REPLACE FUNCTION show_on_segments(show_name text)
+RETURNS table (guc_name text, guc_setting text)
+EXECUTE ON ALL SEGMENTS AS $$
+BEGIN
+	RETURN QUERY SELECT name, setting FROM pg_settings WHERE show_name = name;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Explicitly set GUCs through gpconfig, then run gpstop -u.
+--
+-- log_min_messages has GUC_GPDB_NEED_SYNC in flags, so it should be
+-- synchronized across the segments, and a context of PGC_SUSET, that it should
+-- not be possible to change it without superuser priveleges.
+--
+-- backslash_quote is not synchronized, so it should remain unchanged on
+-- segments.
+
+--start_ignore
+-- Set default values and reconnect.
+\!gpconfig -c log_min_messages -v 'warning'
+\!gpconfig -c backslash_quote -v 'safe_encoding'
+\!gpstop -u
+--end_ignore
+
+-- PGC_SUSET, we don't have enough priveleges.
+RESET log_min_messages;
+-- Successfully reset.
+RESET backslash_quote;
+
+SHOW log_min_messages;
+SELECT * from show_on_segments('log_min_messages');
+
+SHOW backslash_quote;
+SELECT * from show_on_segments('backslash_quote');
+
+--start_ignore
+\!gpconfig -c log_min_messages -v 'warning' -m 'notice'
+\!gpconfig -c backslash_quote -v 'off' -m 'on'
+\!gpstop -u
+--end_ignore
+
+-- Should be 'notice', synchronized with segments.
+SHOW log_min_messages;
+SELECT * from show_on_segments('log_min_messages');
+
+-- Should be 'on' on QD, 'off' on QEs.
+SHOW backslash_quote;
+SELECT * from show_on_segments('backslash_quote');
+
+--start_ignore
+\!gpconfig -r log_min_messages
+\!gpconfig -r backslash_quote
+\!gpstop -u
+--end_ignore
+
+-- After reset, QD and QE should hold the default value, 'warning'.
+SHOW log_min_messages;
+SELECT * from show_on_segments('log_min_messages');
+
+-- After reset, QD and QE should hold the default value, 'safe_encoding'.
+SHOW backslash_quote;
+SELECT * from show_on_segments('backslash_quote');
+
+RESET SESSION AUTHORIZATION;
+
+-- Our session should not have been reset, and this table should still be
+-- present. If it's not, something is terribly wrong and we were silently
+-- disconnected.
+SELECT * FROM dispatch_test_session_not_dead;
```

Comment:
I have the same successful behavior on not patched version. If it's an expected behavior, we possibly don't need this table in tests.

Reply:
> Maybe in case of an error during synchronization we should throw fatal error, or we should abort the transaction and reset the session?

Now, session resets if any of the GUCs fail to be sent. GUCs will be re-sent in the startup packet as usual.

Reply:
> Now, session resets if any of the GUCs fail to be sent. GUCs will be re-sent in the startup packet as usual.

could you add tests for it?

Reply:
> could you add tests for it?

https://github.com/arenadata/gpdb/pull/881#issuecomment-2049822909